libkbfs/bserver_remote.go

```diff
@@ -27,11 +26,18 @@ const (
 	BServerTokenExpireIn = 2 * 60 * 60 // 2 hours
 )
 
+type blockServerRemoteConfig interface {
+	diskBlockCacheGetter
+	codecGetter
+	signerGetter
+	currentInfoGetterGetter
+	logMaker
+}
+
 // BlockServerRemote implements the BlockServer interface and
 // represents a remote KBFS block server.
 type BlockServerRemote struct {
-	codec      kbfscodec.Codec
-	cig        currentInfoGetter
+	config     blockServerRemoteConfig
```

Comment:
I like to think of `Config` as a Factory design pattern that we happen to pass around (except of course we're instantiating all the individual interfaces beforehand). By creating `fooConfig` interfaces, we're in fact specifying which `Config` factories the `foo` component needs, and enforcing a common API for our entire library.

I'm wary of making a distinction between "small" class and "big" class, because "small" classes have a habit of turning into "big" classes over time. By applying different design approaches for the two classification, we're fracturing our design. This leads to confusion when touching new areas of the code:

* Is this a "big" class or a "small" class?
* Should I use the existing `Config` pattern or explicitly name interfaces that I know will be provided by the `Config` interface from upstream?
* What happens if a test subs out the `Config`'s component after we've instantiated ourselves, or worse, if a real component does it (`ResetCaches()`)? Do I need to plumb the updated component through to everything that could be affected by it?

I acknowledge your point about having to think about things when you change something outside of design parameters. But from my perspective, simply requiring that `Config` provides the API you need is enough to think about it. Yes, circular dependencies could be a problem, but there's nothing stopping you from introducing those explicitly either (within the same package).

And then there is the huge timesaving bonus of being able to create one `testFooInterface` type for each `Config` method, and simply passing a new one of those in for every test when we need it.

Regarding the `fooGetter` vs `Foo` embedding conversation, I'm fine either way, but the current `Config` pattern kind of forces our hand there.